src/compiler/checker.ts

```diff
@@ -43062,7 +43104,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     cancellationToken.throwIfCancellationRequested();
             }
         }
-        if (kind >= SyntaxKind.FirstStatement && kind <= SyntaxKind.LastStatement && node.flowNode && !isReachableFlowNode(node.flowNode)) {
+        if (kind >= SyntaxKind.FirstStatement && kind <= SyntaxKind.LastStatement && canHaveFlowNode(node) && node.flowNode && !isReachableFlowNode(node.flowNode)) {

```

Comment:
If you unconditionally access `flowNode` you'll end up with multiple inline cache misses and "wrong map" deoptimizations since the property doesn't exist. Since not all statements have a `flowNode`, we need the check both to satisfy the type system and ensure we're only attempting to access the property on nodes that will have that property present (even if its only been set to `undefined`).

This still isn't the most efficient approach, since accessing `.kind` will almost always be megamorphic. In the long run, we should actually duplicate this code to specific branches in the following `switch` (and ensure the function has only one `switch` on `kind`), but that will come in a later PR. Right now the optimizations I'm applying are broad strokes, I plan to narrow in to specific functions later.